go-controller/pkg/clustermanager/clustermanager_test.go

```diff
@@ -887,6 +888,119 @@ var _ = ginkgo.Describe("Cluster Manager", func() {
 			})
 			gomega.Expect(err).NotTo(gomega.HaveOccurred())
 		})
+
+		ginkgo.It("check for CNC tunnel keys allocations", func() {
+			app.Action = func(_ *cli.Context) error {
+				// CNC uses networkID 4097 (4096+1) which allocates from the idsAllocator range
+				// The idsAllocator starts at 16715779 (16711683 + 4096)
+				// create CNC with already allocated tunnel key
+				cnc1 := &networkconnect.ClusterNetworkConnect{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "cnc1",
+						Annotations: map[string]string{
+							util.OvnConnectRouterTunnelKeyAnnotation: "16715779",
+						},
+					},
+				}
+				// create CNC without tunnel key annotation
+				cnc2 := &networkconnect.ClusterNetworkConnect{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "cnc2",
+					},
+				}
+				clientSet := util.GetOVNClientset(cnc1, cnc2)
+
+				// init the allocator that should reserve already allocated key for cnc1
+				allocator, err := initTunnelKeysAllocator(clientSet.NetworkAttchDefClient, clientSet.NetworkConnectClient)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				// check that reserving different keys for cnc1 will fail
+				err = allocator.ReserveKeys("cnc1", []int{16715780})
+				gomega.Expect(err).To(gomega.HaveOccurred())
+				gomega.Expect(err.Error()).To(gomega.ContainSubstring("can't reserve ids [16715780] for the resource cnc1. It is already allocated with different ids [16715779]"))
+				// now try to allocate key for cnc1 (using networkID 4097 as CNCs do)
+				// and check that returned ID is the already reserved one
+				ids, err := allocator.AllocateKeys("cnc1", 4097, 1)
+				gomega.Expect(err).ToNot(gomega.HaveOccurred())
+				gomega.Expect(ids).To(gomega.Equal([]int{16715779}))
+				// now allocate id for cnc2 (which had no annotation, also using networkID 4097)
+				ids, err = allocator.AllocateKeys("cnc2", 4097, 1)
+				gomega.Expect(err).ToNot(gomega.HaveOccurred())
+				gomega.Expect(ids).To(gomega.Equal([]int{16715780}))
+				// now try cnc3 to make sure IDs of cnc1 and cnc2 are not allocated again
+				ids, err = allocator.AllocateKeys("cnc3", 4097, 1)
+				gomega.Expect(err).ToNot(gomega.HaveOccurred())
+				gomega.Expect(ids).To(gomega.Equal([]int{16715781}))
+				return nil
+			}
+
+			err := app.Run([]string{
+				app.Name,
+				"-cluster-subnets=" + clusterCIDR,
+			})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		})
+
+		ginkgo.It("check for combined NAD and CNC tunnel keys allocations", func() {
```

Comment:
should we add a test for subnet parsing here too? Just to make sure the whole chain works fine

Reply:
yeah I wanted to mostly check the right startup/initial Sync by just running clusterManager. We have unit tests for all functions separately, but I don't think we have a e2e for pre-existing resources (so restart kind of scenario). We could just create and start clusterManager with pre-populated kapi CNC that has e.g. 1 tunnel id and some assigned networks, then make sure it was parsed on startup
not a strong opinion though, if it is hard to implement we can skip

Reply:
spoke with @npinaeva offline, i know what she means.. will do a test in CM with parser.